kebechet/source_management_ogr.py

```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python3
+# Kebechet
+# Copyright(C) 2018, 2019, 2020 Fridolin Pokorny
+#
+# This program is free software: you can redistribute it and / or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Abstract calls to GitHub and GitLab APIs."""
+
+import logging
+import typing
+
+import requests
+from urllib.parse import quote_plus
+
+from ogr.services.github import GithubService
+from ogr.services.gitlab import GitlabService
+
+from .enums import ServiceType
+
+
+_LOGGER = logging.getLogger(__name__)
+BASE_URL = {"github": "https://api.github.com", "gitlab": "https://gitlab.com//api/v4"}
+
+
+class SourceManagement:
+    """Abstract source code management services like GitHub and GitLab."""
+
+    def __init__(
+        self, service_type: ServiceType, service_url: str, token: str, slug: str
+    ):
+        """Initialize source code management tools abstraction.
+
+        Note that we are using OGR for calls. OGR keeps URL to services in its global context per GitHub/GitLab.
+        This is global context is initialized in the manager with a hope to fix this behavior for our needs.
+        """
+        self.service_type = service_type
+        self.slug = slug
+        self.service_url = service_url
+        self.token = token
+        self.namespace, self.repo = slug.rsplit("/", 1)
+
+        if self.service_type == ServiceType.GITHUB:
+            self.service = GithubService(self.token)
+            self.repository = self.service.get_project(
+                repo=self.repo, namespace=self.namespace
+            )
+        elif self.service_type == ServiceType.GITLAB:
+            self.service = GitlabService(self.token)
+            self.repository = self.service.get_project(
+                repo=self.repo, namespace=self.namespace
+            )
+        else:
+            raise NotImplementedError
+
+    def get_issue(self, title: str) -> Issue:
+        """Retrieve issue with the given title."""
+        for issue in self.repository.get_issue_list():
+            if issue._raw_issue.title == title:
+                return issue
```

Comment:
Checkout https://github.com/thoth-station/solver/issues/270 and https://github.com/thoth-station/solver/issues/243 , not sure what the expected behavior would be here.  Should we get the oldest one, the newest one, or both?

Reply:
The newest one.

Reply:
This should get the newest one :+1: